File: Third Year/SEM V/WP-II/LAB/EXP8/webscraper.py
import requests
from bs4 import BeautifulSoup
import csv

# URL of the practice website
url = 'http://books.toscrape.com/'

# Send a GET request to the website
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

# Open a CSV file to store the scraped data
with open('books_data.csv', mode='w', newline='', encoding='utf-8') as file:
    writer = csv.writer(file)
    writer.writerow(['Title', 'Price', 'Availability'])

    # Find all book containers on the page
    books = soup.find_all('article', class_='product_pod')

    for book in books:
        # Extract book title
        title = book.find('h3').find('a')['title']

        # Extract book price
        price = book.find('p', class_='price_color').get_text(strip=True)

        # Extract availability status
        availability = book.find('p', class_='instock availability').get_text(strip=True)

        # Write data to CSV file
        writer.writerow([title, price, availability])

print("Data scraping completed. Check 'books_data.csv' for results.")

File: Third Year/SEM V/WP-II/LAB/EXP8/webscraper.php
<?php
// Set the target URL for the book website
$targetUrl = "http://books.toscrape.com/";
$ch = curl_init();
curl_setopt($ch, CURLOPT_URL, $targetUrl);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

// Execute cURL to fetch HTML content
$htmlContent = curl_exec($ch);
if(curl_errno($ch)) {
    echo 'cURL error: ' . curl_error($ch);
    curl_close($ch);
    exit;
}
curl_close($ch);

// Load HTML content into DOMDocument
$dom = new DOMDocument();
libxml_use_internal_errors(true); // Suppress warnings for invalid HTML
$dom->loadHTML($htmlContent);
libxml_clear_errors();

// Use XPath to locate book information
$xpath = new DOMXPath($dom);
$books = $xpath->query("//article[@class='product_pod']");

// Open a CSV file to save the scraped data
$file = fopen("books_data.csv", "w");
fputcsv($file, ['Title', 'Price', 'Availability']); // Header row

// Iterate through each book container and extract details
foreach ($books as $book) {
    // Extract book title
    $titleNode = $xpath->query(".//h3/a", $book)->item(0);
    $title = $titleNode ? $titleNode->getAttribute("title") : "N/A";

    // Extract book price
    $priceNode = $xpath->query(".//p[@class='price_color']", $book)->item(0);
    $price = $priceNode ? $priceNode->textContent : "N/A";

    // Extract availability status
    $availabilityNode = $xpath->query(".//p[contains(@class, 'instock')]", $book)->item(0);
    $availability = $availabilityNode ? trim($availabilityNode->textContent) : "N/A";

    // Write data to CSV file
    fputcsv($file, [$title, $price, $availability]);
}

// Close the CSV file
fclose($file);

echo "Data scraping completed. Check 'books_data.csv' for results.";
?>
